File: pages/producto_page.ts
import { Page, Locator, expect } from '@playwright/test';
import config from '../config/config';

export class ProductoFormPage {
  readonly page: Page;
  readonly comboboxAnio: Locator;
  readonly opcion2038: Locator;
  readonly observaciones: Locator;
  readonly aseguradoInput: Locator;
  readonly aseguradoOption: Locator;
  readonly productoInput: Locator;
  readonly mesInput: Locator;

  constructor(page: Page) {
    this.page = page;
    this.comboboxAnio = page.locator('button[aria-label*="Año"]');
    this.opcion2038 = page.locator('lightning-base-combobox-item[data-value="2038"]');
    this.observaciones = page.locator('input').nth(1);
    this.aseguradoInput = page.locator('input[role="combobox"]').nth(2);
    this.aseguradoOption = page.locator('text=5.000.000');
    this.productoInput = page.locator('input[role="combobox"]').first();
    this.mesInput = page.locator('input[role="combobox"]').nth(1);
  }

  async datosProducto() {
    await this.comboboxAnio.scrollIntoViewIfNeeded();
    await this.comboboxAnio.click({ timeout: 10000 });

    await this.opcion2038.waitFor({ state: 'visible', timeout: 10000 });
    await this.opcion2038.click();

    await this.observaciones.waitFor({ state: 'visible', timeout: 10000 });
    await this.observaciones.click();
    await this.observaciones.fill('Ninguno');

    console.log('Año de ingreso a cotizar');
  }

async listaAsegurado() {
    await this.aseguradoInput.scrollIntoViewIfNeeded();
    await this.aseguradoInput.click();
    await this.aseguradoInput.fill(config.VALOR_ASEGURADO || '1.000.000');

    await this.aseguradoOption.waitFor({ state: 'visible', timeout: 10000 });
    await this.aseguradoOption.click();
  }

  async datosProductoSeguraPlus() {
    // PRODUCTO
    await this.productoInput.scrollIntoViewIfNeeded();
    await this.productoInput.click();

    // Simular navegación con flechas
    await this.page.keyboard.press('ArrowDown');
    await this.page.keyboard.press('Enter');

    // MES
    await this.mesInput.scrollIntoViewIfNeeded();
    await this.mesInput.click();
    await this.page.keyboard.press('Enter');

    // ASEGURADO
    await this.listaAsegurado();
  }
  async datosProductoSeguraPlusSemestre() {
    // PRODUCTO
    await this.productoInput.scrollIntoViewIfNeeded();
    await this.productoInput.click();
    await this.page.waitForTimeout(500); // breve espera para evitar errores
  
    await this.page.keyboard.press('ArrowDown'); // para bajar una vez
    await this.page.keyboard.press('Enter');
  
    // MES
    await this.mesInput.scrollIntoViewIfNeeded();
    await this.mesInput.click();
    await this.page.keyboard.press('Enter');
  
    // ASEGURADO
    await this.aseguradoInput.scrollIntoViewIfNeeded();
    await this.aseguradoInput.click();
    await this.aseguradoInput.fill(config.VALOR_ASEGURADO || '1.000.000');
    await this.aseguradoOption.waitFor({ state: 'visible' });
    await this.aseguradoOption.click();
  
    // SEMESTRE
    const semestreInput = this.page.locator(
      'vlocity_ins-omniscript-select >> nth=15 input'
    );
    await semestreInput.scrollIntoViewIfNeeded();
    await semestreInput.click();
    for (let i = 0; i < 6; i++) {
      await this.page.keyboard.press('ArrowDown');
      await this.page.waitForTimeout(300);
    }
    await this.page.keyboard.press('Enter');
    console.log('Opción semestre 6 seleccionada');
  }
  async datosProductoGmProfesional() {
    await this.page.waitForTimeout(5000);
    const productoElemento = this.page.locator('input[role="combobox"]');
    await productoElemento.scrollIntoViewIfNeeded();
    await productoElemento.click();
    await productoElemento.press('ArrowDown');
    await productoElemento.press('Enter');

    const tarifa = this.page.locator('xpath=/html/body/.../input');
    await tarifa.click();
    await tarifa.fill(process.env.MES_TARIFA || '');
    await tarifa.press('Enter');
    await this.page.locator('xpath=/html/body/.../span[text()="...opcion..."]').click();

    const asegurado = this.page.locator('xpath=/html/body/.../input');
    await asegurado.click();
    await asegurado.fill(process.env.VALOR_ASEGURADO2 || '');
    await this.page.locator('xpath=//span[text()="2.000.000"]').click();

    const año = this.page.locator('xpath=/html/body/.../button');
    await año.click();
    for (let i = 0; i < 6; i++) await año.press('ArrowDown');
    await año.press('Enter');

    const periodo = this.page.locator('xpath=/html/body/.../input');
    await periodo.click();
    for (let i = 0; i < 6; i++) await periodo.press('ArrowDown');
    await periodo.press('Enter');
  }

  async simuladorProductoSeguraPlus() {
    const titulo = this.page.locator('text=Valor asegurado semestre');
    await expect(titulo).toBeVisible();
    await this.page.waitForTimeout(10000);
    console.log("El texto '$ 0,00' está presente en la página.");
  }

  async simuladorProductoSeguraPlusSemestre() {
    const titulo = this.page.locator('text=GlobalUniversidad Segura Plus Semestres');
    await expect(titulo).toBeVisible();
    const combobox = this.page.locator('input[role="combobox"]');
    await combobox.click();
    await this.page.locator('text=3.000.000').click();
    await this.page.waitForTimeout(5000);
    console.log("El texto '$ 14.538.000,00' está presente en la página.");
  }

  async simuladorProductoGmProfesional() {
    const combobox = this.page.locator('input[role="combobox"]');
    await combobox.click();
    await this.page.locator('text=3.000.000').click();

    const segundoCombo = this.page.locator('xpath=/html/body/.../input');
    await segundoCombo.click();
    await this.page.locator('text=3').click();

    const tercerCombo = this.page.locator('xpath=/html/body/.../input');
    await tercerCombo.click();
    await this.page.locator('text=Trimestral').click();
  }

  async buttonCotizar() {
    const boton = this.page.locator('button:has-text("Cotizar")');
    await boton.scrollIntoViewIfNeeded();
    await boton.click();
    await this.page.waitForTimeout(20000);
  }

  async botonAtras() {
    const boton = this.page.locator('button:has-text("Atras")');
    await boton.scrollIntoViewIfNeeded();
    await boton.click();
  }

  async botonAnterior() {
    const boton = this.page.locator('button:has-text("Atras")');
    await boton.scrollIntoViewIfNeeded();
    await boton.click();
  }
}

File: pages/producto_page.py
# pages/producto_page.py
from pages.base_page import BasePage
from selenium.webdriver.common.by import By
from selenium.webdriver.support.ui import WebDriverWait
from selenium.webdriver.support import expected_conditions as EC
from selenium.webdriver.common.keys import Keys
import time


class ProductoFormPage(BasePage):
    def __init__(self, driver):
        self.driver = driver
        self.wait = WebDriverWait(driver, 30)

    # ==============================
    # AÑO / OBSERVACIONES
    # ==============================
    def datos_producto(self):
        w = WebDriverWait(self.driver, 25)

        # Año
        anio_btn = w.until(EC.element_to_be_clickable((
            By.XPATH, "//button[@role='combobox' and @aria-label='Año de ingreso a cotizar']"
        )))
        self.driver.execute_script("arguments[0].scrollIntoView({block:'center'});", anio_btn)
        anio_btn.click()
        w.until(EC.presence_of_element_located((
            By.XPATH, "//div[@role='listbox' and contains(@id,'dropdown-element')]"
        )))
        opcion_2038 = w.until(EC.element_to_be_clickable((
            By.XPATH, "//lightning-base-combobox-item[@data-value='2038']"
        )))
        self.driver.execute_script("arguments[0].click();", opcion_2038)
        time.sleep(0.3)

        # Observaciones
        obs = w.until(EC.element_to_be_clickable((
            By.XPATH, "//label[normalize-space()='Observaciones']/following::input[1]"
        )))
        self.driver.execute_script("arguments[0].scrollIntoView({block:'center'});", obs)
        obs.clear()
        obs.send_keys("Ninguno")

    # ==============================
    # HELPERS PRODUCTO
    # ==============================
    def _find_producto_wrapper(self):
        # Preferido por data-omni-key
        try:
            wrap = self.driver.find_element(By.XPATH, "//*[starts-with(@data-omni-key,'ProductoSeleccionado')]")
            self.driver.execute_script("arguments[0].scrollIntoView({block:'center'});", wrap)
            return wrap
        except Exception:
            pass
        # Fallback: segundo select (similar a grabación)
        try:
            wrap = self.driver.find_element(By.XPATH, "(//vlocity_ins-omniscript-select)[2]")
            self.driver.execute_script("arguments[0].scrollIntoView({block:'center'});", wrap)
            return wrap
        except Exception:
            return None

    def _get_producto_container_and_control(self, wrapper):
        container = wrapper.find_element(By.XPATH, ".//div[contains(@class,'slds-combobox')][1]")
        control = wrapper.find_element(
            By.XPATH,
            ".//div[contains(@class,'slds-form-element__control')]//div[contains(@class,'slds-combobox__form-element')]"
        )
        return container, control

    def _abrir_producto(self):
        wrap = self._find_producto_wrapper()
        if not wrap:
            raise RuntimeError("No se encontró el componente de Producto.")
        container, control = self._get_producto_container_and_control(wrap)
        try:
            inp = container.find_element(By.XPATH, ".//input[@role='combobox']")
        except Exception:
            inp = None

        for _ in range(3):
            if inp:
                self.driver.execute_script("arguments[0].scrollIntoView({block:'center'});", inp)
                self.driver.execute_script("arguments[0].click();", inp)
                time.sleep(0.15)
            self.driver.execute_script("arguments[0].scrollIntoView({block:'center'});", control)
            self.driver.execute_script("arguments[0].click();", control)
            time.sleep(0.2)
            cls = container.get_attribute("class") or ""
            expanded = (inp.get_attribute("aria-expanded") if inp else "") or ""
            if "slds-is-open" in cls or expanded.lower() == "true":
                break

        listbox = container.find_element(By.XPATH, ".//div[@role='listbox']")
        return container, listbox

    def _click_opcion(self, listbox, label_text):
        try:
            op = listbox.find_element(By.XPATH, f".//div[@role='option' and @data-label=\"{label_text}\"]")
        except Exception:
            op = listbox.find_element(By.XPATH, f".//span[normalize-space()=\"{label_text}\"]/ancestor::div[@role='option']")
        # eventos nativos
        self.driver.execute_script("""
            const el = arguments[0];
            for (const t of ['pointerdown','mousedown','mouseup','click']) {
                el.dispatchEvent(new MouseEvent(t,{bubbles:true,cancelable:true,composed:true}));
            }
        """, op)
        return op

    def _hard_set_producto(self, container, desired_label, option_elem=None):
        """
        Plan B: setear el valor por JS + disparar eventos para que OmniScript haga binding.
        """
        self.driver.execute_script("""
            const container = arguments[0];
            const desired = arguments[1];
            const option = arguments[2];
            const inp = container.querySelector("input[role='combobox']");
            let dv = "";
            if (option) {
                dv = option.getAttribute("data-value") || "";
            } else {
                const lb = container.querySelector("div[role='listbox']");
                const found = Array.from(lb.querySelectorAll("div[role='option']")).find(
                  o => (o.getAttribute('data-label')===desired) || (o.textContent.trim()===desired)
                );
                if (found) dv = found.getAttribute("data-value") || "";
            }
            if (inp) {
                inp.value = desired;
                inp.setAttribute("data-value", dv);
                inp.setAttribute("aria-invalid", "false");
                inp.dispatchEvent(new Event("input", {bubbles:true}));
                inp.dispatchEvent(new Event("change", {bubbles:true}));
                inp.dispatchEvent(new Event("blur", {bubbles:true}));
            }
        """, container, desired_label, option_elem)

    def _confirm_producto_set(self, container, desired_label, timeout=6):
        end = time.time() + timeout
        try:
            inp = container.find_element(By.XPATH, ".//input[@role='combobox']")
        except Exception:
            inp = None
        last_val = ""
        while time.time() < end and inp:
            val = (inp.get_attribute("value") or "").strip()
            ai = (inp.get_attribute("aria-invalid") or "").strip().lower()
            last_val = val
            if val == desired_label or ai == "false":
                return True
            time.sleep(0.25)
        return False

    def _seleccionar_producto(self, etiqueta_visible):
        container, listbox = self._abrir_producto()

        # 1) intento normal con click a la opción
        opcion = self._click_opcion(listbox, etiqueta_visible)
        if self._confirm_producto_set(container, etiqueta_visible, timeout=3):
            return

        # 2) HARD SET por JS (forzar binding)
        self._hard_set_producto(container, etiqueta_visible, opcion)
        if self._confirm_producto_set(container, etiqueta_visible, timeout=3):
            return

        # 3) Blur con TAB y reintento de confirmación
        try:
            inp = container.find_element(By.XPATH, ".//input[@role='combobox']")
            inp.send_keys(Keys.TAB)
            time.sleep(0.3)
        except Exception:
            pass

        if not self._confirm_producto_set(container, etiqueta_visible, timeout=2):
            try:
                ai = inp.get_attribute("aria-invalid") if inp else "N/A"
                val = (inp.get_attribute("value") if inp else "N/A")
            except Exception:
                ai, val = "N/A", "N/A"
            raise AssertionError(f"Producto no quedó seleccionado. value='{val}', aria-invalid={ai}")

    # ==============================
    # FLUJOS DE PRODUCTO
    # ==============================
    def datos_producto_segura_plus(self):
        self._seleccionar_producto("Global Universidad Segura Plus")
        self._llenar_campos_comunes()

    def datos_producto_segura_plus_semestre(self):
        self._seleccionar_producto("Global Universidad Segura Plus Semestres")
        self._llenar_campos_comunes()

    def datos_producto_gmprofesional(self):
        self._seleccionar_producto("GlobalMás Profesional")
        self._llenar_campos_comunes()

    # ==============================
    # CAMPOS COMUNES (Tarifa)
    # ==============================
    def _llenar_campos_comunes(self):
        try:
            self.seleccionar_tarifa_septiembre()
        except Exception as e:
            print(f"⚠ No se pudo seleccionar tarifa: {e}")

    def seleccionar_tarifa_septiembre(self, timeout=15):
        w = WebDriverWait(self.driver, timeout)
        cont = w.until(EC.presence_of_element_located((
            By.XPATH, "//label[.//span[normalize-space()='Tarifa']]/ancestor::div[contains(@class,'slds-combobox')][1]"
        )))
        try:
            opener = cont.find_element(By.XPATH, ".//input[@role='combobox']")
        except Exception:
            opener = cont.find_element(By.XPATH, ".//span[contains(@class,'slds-input__icon_right')]")
        self.driver.execute_script("arguments[0].scrollIntoView({block:'center'});", opener)
        self.driver.execute_script("arguments[0].click();", opener)

        # listbox (aunque no marque slds-is-open)
        lb = cont.find_element(By.XPATH, ".//div[@role='listbox']")

        # intenta “Septiembre”, si no existe, toma el primer option disponible
        opcion = None
        try:
            opcion = lb.find_element(By.XPATH, ".//div[@role='option' and @data-label='Septiembre']")
        except Exception:
            try:
                opcion = lb.find_element(By.XPATH, ".//span[normalize-space()='Septiembre']/ancestor::div[@role='option']")
            except Exception:
                # primer opción disponible
                opcion = lb.find_element(By.XPATH, ".//div[@role='option'][1]")

        self.driver.execute_script("arguments[0].click();", opcion)
        print("✅ Tarifa seleccionada")

    # ==============================
    # SIMULADORES (validaciones)
    # ==============================
    def simulador_producto_segura_plus(self):
        # si te sigue fallando aquí, probablemente no quedó el producto → ver arriba
        titulo = WebDriverWait(self.driver, 20).until(
            EC.visibility_of_element_located((By.XPATH, "//*[normalize-space()='Valor asegurado semestre']"))
        )
        assert titulo.is_displayed()

    def simulador_producto_segura_plus_semestre(self):
        titulo = WebDriverWait(self.driver, 20).until(
            EC.visibility_of_element_located((By.XPATH, "//*[normalize-space()='Global Universidad Segura Plus Semestres']"))
        )
        assert titulo.is_displayed()

    def simulador_producto_gmprofesional(self):
        titulo = WebDriverWait(self.driver, 20).until(
            EC.visibility_of_element_located((By.XPATH, "//*[contains(normalize-space(),'GlobalMás Profesional')]"))
        )
        assert titulo.is_displayed()

    # ==============================
    # BOTONES
    # ==============================
    def button_cotizar(self):
        # Sólo dispara si el campo Producto NO está en error
        try:
            wrap = self._find_producto_wrapper()
            cont, _ = self._get_producto_container_and_control(wrap)
            inp = cont.find_element(By.XPATH, ".//input[@role='combobox']")
            ai = (inp.get_attribute("aria-invalid") or "").lower()
            if ai == "true":
                raise AssertionError("Producto sigue requerido. No se puede Cotizar.")
        except Exception:
            # si no podemos leer el estado, continuamos, pero el simulador puede fallar
            pass

        btn = self.wait.until(EC.element_to_be_clickable((By.XPATH, "//button[span[normalize-space()='Cotizar']]")))
        self.driver.execute_script("arguments[0].scrollIntoView({block:'center'});", btn)
        btn.click()
        time.sleep(1.5)

    def boton_anterior(self):
        btn = self.wait.until(EC.element_to_be_clickable((By.XPATH, "//button[span[normalize-space()='Anterior']]")))
        self.driver.execute_script("arguments[0].scrollIntoView({block:'center'});", btn)
        btn.click()
        time.sleep(1.0)

    def button_finalizar(self):
        btn = self.wait.until(EC.element_to_be_clickable((By.XPATH, "//button[span[normalize-space()='Finalizar']]")))
        self.driver.execute_script("arguments[0].scrollIntoView({block:'center'});", btn)
        btn.click()
        time.sleep(1.5)
